comb_spec_searcher/rule_db/forest.py

```diff
@@ -0,0 +1,638 @@
+            try:
+                res[forest_key.bucket].append(forest_key)
+            except KeyError as e:
+                msg = (
+                    f"{forest_key.bucket} type is not currently supported "
+                    "by the extractor"
+                )
+                raise RuntimeError(msg) from e
+        return res
+
+    def _find_rule(self, rule_key: ForestRuleKey) -> AbstractRule:
+        """
+        Find a rule that have the given rule key.
+        """
+        all_classes = (rule_key.parent,) + rule_key.children
+        all_normal_rules = itertools.chain.from_iterable(
+            self._rules_for_class(c) for c in all_classes
+        )
+        for normal_rule in all_normal_rules:
+            potential_rules = [normal_rule]
+            if normal_rule.is_reversible():
+                assert isinstance(normal_rule, Rule)
+                potential_rules.extend(
+                    normal_rule.to_reverse_rule(i)
+                    for i in range(len(normal_rule.children))
+                )
+            for rule in potential_rules:
+                if rule.forest_key(self.classdb.get_label) == rule_key:
+                    return rule
+        err = f"Can't find a rule for {rule_key}\n"
+        err += f"Parent:\n{self.classdb.get_class(rule_key.parent)}\n"
+        for i, l in enumerate(rule_key.children):
+            err += f"Child {i}:\n{self.classdb.get_class(l)}\n"
+        raise RuntimeError(err)
+
+    def _rules_for_class(self, label: int) -> Iterator[AbstractRule]:
+        """
+        Return all the rule created for that class with the pack.
+        """
+        comb_class = self.classdb.get_class(label)
+        strats: Iterator[CSSstrategy] = itertools.chain(self.pack, [EmptyStrategy()])
+        for strat in strats:
+            if isinstance(strat, StrategyFactory):
+                strats_or_rules: Iterable[
+                    Union[AbstractRule, AbstractStrategy]
+                ] = strat(comb_class)
+            else:
+                strats_or_rules = [strat]
+            for x in strats_or_rules:
+                if isinstance(x, AbstractStrategy):
+                    try:
+                        yield x(comb_class)
+                    except StrategyDoesNotApply:
+                        continue
+                else:
+                    yield x
+
+
+class RuleDBForest(RuleDBAbstract):
+    """
+    The rule database that provides live information on which class are pumping with the
+    current rule in the database.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._num_rules = 0
+        self.table_method = TableMethod()
+        self._already_empty: Set[int] = set()
+
+    # Implementation of RuleDBAbstract
+
+    def status(self, elaborate: bool) -> str:
+        s = "RuleDB status:\n"
+        s += f"\tAdded from {self._num_rules} normal rules\n"
+        s += self.table_method.status()
```

Comment:
Could we add some timings, such as how long is spent processing the queue, so we can keep track of the speed. This is currently all hidden in the strategy timings in CSS.

Reply:
Fixed